libpermian/plugins/anaconda_webui/__init__.py

```diff
@@ -0,0 +1,443 @@
+            location = f'{os_url},kernel={self.installation_source.kernel_path},initrd={self.installation_source.initrd_path}'
+        else:
+            location = os_url
+
+        # Construct kernel cmdline for --extra-args
+        kernel_cmdline = f'inst.sshd inst.webui inst.webui.remote inst.stage2={self.os_url}'
+        if self.additional_repos:
+            for repo in self.additional_repos:
+                kernel_cmdline += f' inst.addrepo={self.installation_source.repo[repo][self.architecture]}'
+
+        if self.kernel_cmdline:
+            kernel_cmdline += f' {self.kernel_cmdline}'
+
+        # Assemble virt-install command
+        cmd = ['virt-install', '--connect', self.hypervisor_host, '--noautoconsole',
+            '-n', self.vm_name, '--os-variant', 'rhel-unknown', '--location', location,
+            '--memory', '4096', '--vcpus', '2', '--disk', 'size=10',
+            '--extra-args', kernel_cmdline]
+
+        LOGGER.info('Running: ' + repr(cmd))
+        process = subprocess.run(cmd ,check=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+        self.log(process.stdout.decode())
+
+        # Wait for machine to start and get IP
+        self.log('Waiting for IP', show=True)
+        startup_timeout = 120 + time.time()
+        while True:
+            if startup_timeout < time.time():
+                Exception('Timeout, VM still doesn\'t have IP')
+            time.sleep(10)
+            output = self._virsh_call(['domifaddr', self.vm_name])
+            if output:
+                break
+
+        return output.split()[3].split('/')[0]
+
+    def _virsh_call(self, args, check=False):
+        """ Runs virsh with specified arguments
+
+        :param args: virsh arguments
+        :type args: list
+        :param check: Check if command was successful, defaults to False
+        :type check: bool, optional
+        :return: CompletedProcess
+        :rtype: subprocess.CompletedProcess
+        """
+        cmd = ['virsh', '-q', '--connect', self.hypervisor_host] + args
+        LOGGER.debug('Running: ' + repr(cmd))
+        return subprocess.run(cmd, check=check, stdout=subprocess.PIPE).stdout.decode()
+
+    def _clone_repo(self, repo, branch, dir):
+        """ Clones git repository
+
+        :param repo: Repository URL
+        :type repo: str
+        :param branch: Branch name
+        :type branch: str
+        :param dir: Directory where the repo should be cloned
+        :type dir: str
+        """
+        git_cmd = ['git', 'clone', '--depth', '1', '-b', branch, repo, dir]
+        LOGGER.debug('Running: ' + repr(git_cmd))
+        subprocess.check_call(git_cmd)
+
+    def _get_test_repo_dir(self):
+        """ Clones test repository with the same branch used for anaonda repo, if test repo doesn't already exist """
+        self.test_repo_dir = os.path.join(self.temp_dir, self.test_repo_name)
+        if  not os.path.isdir(self.test_repo_dir):
```

Comment:
This condition is problematic. Due to:

1. it's not thread safe (which may not be a big deal)
2. the condition itself doesn't ensure the desired effect.

The second issue is a bigger problem as the behaviour of this method is "strange":

- When it's empty directory, the function happily finishes
- When it's file, socket, symlink pointing to file, broken symlink or anything other existing than directory, the method attempts to to clone to it (raising traceback which is a good thing in this context)

I cannot properly describe how this individual method could be fixed, it's just strange, also the "get" in the name is misleading as it sounds like the test_repo_dir (path to it) should be returned. Change of name to something like fetch/clone would help this but I still have a bad feeling about it which may signal something wrong in the code structure.

Reply:
This should always be called only from inside setup_lock, so the method itself doesn't have to be thread safe. 
If there would be a problem with the existing directory, it's because setup of different workflow instance failed. That is more general problem of shared setup between workflows.

Reply:
Ok, thanks for the explanation. I'm just thinking if it wouldn't be worth somehow marking such methods that operate on top of shared resource and are not thread safe, but that' just my feeling that it could be good to do something like that. Marking this conversation as resolved.